File: backend/src/db/migrate.ts
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { env } from '../config/env'
import { pool, SCHEMA, withConnectionRetry } from './pool'

const dir = path.join(path.dirname(fileURLToPath(import.meta.url)), 'migrations')

/** Drops every table in the Annex HR schema (the schema itself and all other schemas are untouched). */
async function reset() {
  if (env.isProd) throw new Error('Refusing to reset the database in production.')
  const { rows } = await pool.query<{ tablename: string }>('SELECT tablename FROM pg_tables WHERE schemaname = $1', [env.DB_SCHEMA])
  if (rows.length) {
    await pool.query(`DROP TABLE ${rows.map((r) => `${SCHEMA}."${r.tablename}"`).join(', ')} CASCADE`)
  }
  console.log(`↺ Dropped ${rows.length} tables from schema ${SCHEMA}`)
}

export async function migrate() {
  const exists = await withConnectionRetry(() => pool.query('SELECT 1 FROM pg_namespace WHERE nspname = $1', [env.DB_SCHEMA]))
  if (!exists.rowCount) throw new Error(`Schema ${SCHEMA} does not exist. Create it first: CREATE SCHEMA ${SCHEMA};`)

  await pool.query(`CREATE TABLE IF NOT EXISTS ${SCHEMA}.schema_migrations (name text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())`)
  const applied = new Set((await pool.query<{ name: string }>(`SELECT name FROM ${SCHEMA}.schema_migrations`)).rows.map((r) => r.name))
  const files = fs.readdirSync(dir).filter((f) => f.endsWith('.sql')).sort()

  let count = 0
  for (const file of files) {
    if (applied.has(file)) continue
    const sql = fs.readFileSync(path.join(dir, file), 'utf8')
    const client = await withConnectionRetry(() => pool.connect())
    try {
      await client.query('BEGIN')
      await client.query(`SET LOCAL search_path TO ${SCHEMA}`)
      await client.query(sql)
      await client.query(`INSERT INTO ${SCHEMA}.schema_migrations (name) VALUES ($1)`, [file])
      await client.query('COMMIT')
      console.log(`✓ Applied ${file}`)
      count++
    } catch (err) {
      await client.query('ROLLBACK')
      throw new Error(`Migration ${file} failed: ${(err as Error).message}`)
    } finally {
      client.release()
    }
  }
  console.log(count ? `Migrations complete (${count} applied) in schema ${SCHEMA}.` : `Schema ${SCHEMA} is up to date.`)
}

const isMain = process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)
if (isMain) {
  try {
    if (process.argv.includes('--reset')) await reset()
    await migrate()
  } catch (err) {
    console.error('✖', (err as Error).message)
    process.exitCode = 1
  } finally {
    await pool.end()
  }
}

File: backend/db/migrate.py
"""Migration runner.

    python -m db.migrate            apply pending migrations
    python -m db.migrate --reset    drop every table in DB_SCHEMA first (refused in production)

SQL files in db/migrations run in filename order, each in a transaction with
search_path pinned to DB_SCHEMA, and are recorded in schema_migrations.
"""

from __future__ import annotations

import sys
from pathlib import Path

from api.config import settings
from api.db import SCHEMA, connection, pool, query, with_connection_retry

MIGRATIONS = Path(__file__).resolve().parent / "migrations"


def reset() -> None:
    if settings.is_prod:
        raise SystemExit("Refusing to reset the database in production.")
    rows = query("SELECT tablename FROM pg_tables WHERE schemaname = $1", [settings.DB_SCHEMA])
    if rows:
        query("DROP TABLE " + ", ".join(f'{SCHEMA}."{r["tablename"]}"' for r in rows) + " CASCADE")
    print(f"↺ Dropped {len(rows)} tables from schema {SCHEMA}")


def migrate() -> None:
    if not query("SELECT 1 FROM pg_namespace WHERE nspname = $1", [settings.DB_SCHEMA]):
        raise SystemExit(f"Schema {SCHEMA} does not exist. Create it first: CREATE SCHEMA {SCHEMA};")
    query(f"CREATE TABLE IF NOT EXISTS {SCHEMA}.schema_migrations (name text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())")
    applied = {r["name"] for r in query(f"SELECT name FROM {SCHEMA}.schema_migrations")}
    count = 0
    for file in sorted(MIGRATIONS.glob("*.sql")):
        if file.name in applied:
            continue
        sql = file.read_text(encoding="utf-8")
        with connection() as conn:
            with conn.transaction():
                conn.execute(f"SET LOCAL search_path TO {SCHEMA}")
                conn.execute(sql)  # type: ignore[arg-type] — multi-statement script, no parameters
                conn.execute(f"INSERT INTO {SCHEMA}.schema_migrations (name) VALUES (%s)", [file.name])
        print(f"✓ Applied {file.name}")
        count += 1
    print(f"Migrations complete ({count} applied) in schema {SCHEMA}." if count else f"Schema {SCHEMA} is up to date.")


def main() -> None:
    pool.open()
    try:
        if "--reset" in sys.argv:
            with_connection_retry(reset)
        migrate()
    finally:
        pool.close()


if __name__ == "__main__":
    main()
